frame/session/src/historical/offchain.rs

```diff
@@ -0,0 +1,272 @@
+	pub fn as_slice(&self) -> &[(T::ValidatorId, T::FullIdentification)] {
+		self.validator_set.as_slice()
+	}
+
+	/// Convert `self` into a vector and consume `self`.
+	pub fn into_vec(self) -> Vec<(T::ValidatorId, T::FullIdentification)> {
+		self.validator_set
+	}
+
+	/// Attempt to prune anything that is older than `first_to_keep` session index.
+	///
+	/// Due to re-organisation it could be that the `first_to_keep` might be less
+	/// than the stored one, in which case the conservative choice is made to keep records
+	/// up to the one that is the lesser.
+	pub fn prune_older_than(first_to_keep: SessionIndex) {
+		let derived_key = shared::LAST_PRUNE.to_vec();
+		let entry = StorageValueRef::persistent(derived_key.as_ref());
+		match entry.mutate(|current: Option<Option<SessionIndex>>| -> Result<_, ()> {
+			match current {
+				Some(Some(current)) if current < first_to_keep => Ok(first_to_keep),
+				// do not move the cursor, if the new one would be behind ours
+				Some(Some(current)) => Ok(current),
+				None => Ok(first_to_keep),
+				// if the storage contains undecodable data, overwrite with current anyways
+				// which might leak some entries being never purged, but that is acceptable
+				// in this context
+				Some(None) => Ok(first_to_keep),
+			}
+		}) {
+			Ok(Ok(new_value)) => {
+				// on a re-org this is not necessarily true, with the above they might be equal
+				if new_value < first_to_keep {
+					for session_index in new_value..first_to_keep {
+						let derived_key = shared::derive_key(shared::PREFIX, session_index);
+						let _ = StorageValueRef::persistent(derived_key.as_ref()).clear();
+					}
+				}
+			}
+			Ok(Err(_)) => {} // failed to store the value calculated with the given closure
+			Err(_) => {}     // failed to calculate the value to store with the given closure
+		}
+	}
+
+	/// Keep the newest `n` items, and prune all items older than that.
+	pub fn keep_newest(n_to_keep: usize) {
+		let session_index = <SessionModule<T>>::current_index();
+		let n_to_keep = n_to_keep as SessionIndex;
+		if n_to_keep < session_index {
+			Self::prune_older_than(session_index - n_to_keep)
+		}
+	}
+
+	#[inline]
+	fn len(&self) -> usize {
+		self.validator_set.len()
+	}
+}
+
+/// Implement conversion into iterator for usage
+/// with [ProvingTrie](super::ProvingTrie::generate_for).
+impl<T: Trait> core::iter::IntoIterator for ValidatorSet<T> {
+	type Item = (T::ValidatorId, T::FullIdentification);
+	type IntoIter = sp_std::vec::IntoIter<Self::Item>;
+	fn into_iter(self) -> Self::IntoIter {
+		self.validator_set.into_iter()
+	}
+}
+
+/// Create a proof based on the data available in the off-chain database.
+///
+/// Based on the yielded `MembershipProof` the implementer may decide what
+/// to do, i.e. in case of a failed proof, enqueue a transaction back on
+/// chain reflecting that, with all its consequences such as i.e. slashing.
+pub fn prove_session_membership<T: Trait, D: AsRef<[u8]>>(
```

Comment:
Why not group that into `ValidatorSet` static methods too? I don't think I'm quite following the reasoning behind having some static methods in `ValidatorSet` and some not.

Afaict the whole `ValidatorSet` is only used in this module, maybe best to make it private and only expose the API that is actually necessary outside (which I believe would be this particular function).

Reply:
I think that is the best solution to make the `ValidatorSet` private.